src/librustc_typeck/check/closure.rs

```diff
@@ -40,19 +45,62 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         self.check_closure(expr, expected_kind, decl, body, expected_sig)
     }
 
+    fn declare_closure(&self, def_id: DefId) {
+        let tcx = self.tcx.global_tcx();
+
+        tcx.generics.memoize(def_id, || {
+            let node_id = tcx.map.as_local_node_id(def_id).unwrap();
+            let base_def_id = self.tcx.closure_base_def_id(def_id);
+            let base_generics = tcx.item_generics(base_def_id);
+
+            // provide junk type parameter defs - the only place that
+            // cares about anything but the length is instantiation,
+            // and we don't do that for closures.
+            let upvar_decls : Vec<_> = tcx.with_freevars(node_id, |fv| {
+                fv.iter().enumerate().map(|(i, _)| ty::TypeParameterDef {
+                    index: (base_generics.count() as u32) + (i as u32),
+                    name: token::intern("<upvar>"),
+                    def_id: def_id,
+                    default_def_id: base_def_id,
+                    default: None,
+                    object_lifetime_default: ty::ObjectLifetimeDefault::BaseDefault,
+                    pure_wrt_drop: false,
+                }).collect()
+            });
+
+            tcx.alloc_generics(ty::Generics {
+                parent: Some(base_def_id),
+                parent_regions: base_generics.parent_regions +
+                    (base_generics.regions.len() as u32),
+                parent_types: base_generics.parent_types +
+                    (base_generics.types.len() as u32),
+                regions: vec![],
+                types: upvar_decls,
+                has_self: false,
```

Comment:
This should be `base_generics.has_self`. Nevertheless, putting generics/type *here* is a problem.
In #37676 I've [added generics for closures in `typeck::collect`](https://github.com/rust-lang/rust/pull/37676/commits/68c898bfc2c1efdcfd1b0eabdf8217415a7af77e#diff-4ed25c00aceb84666fca639cf8101c7cR137), which is closer to what on-demand type-checking needs.
OTOH your `closure_base_def_id` is better (or even actually correct) than what I have there, so I'd prefer it.